content-creators-kickstarter/exercise-structure.md

```diff
@@ -0,0 +1,137 @@
+# Content of an exercise repository
+
+## General structure
+
+### Tutorial and Fix
+The solution checking is integrated into the TFW, since all we have to do is check if the state machine is in the last state, so there is no separate `controller` folder. However, stepping the state machine based on user actions is the your task.
+
+The general file structure looks like this:
+```
+├── 1. tfw  
+│   └── [the TFW itself, please don't modify it]  
+├── 2. solvable  
+│   ├── 3. Dockerfile   
+│   ├── 4. nginx    
+│   │   └── [Nginx config files]
+│   ├── 5. supervisor
+│   │   └── [Supervisor config files]
+│   ├── 6. tutorial  
+│   │   ├── 7. app.yml: The YAML file that describes the whole exercise, including steps, views and chatbot messages   
+│   │   └── 8. A file that implements your event handlers using your SDK of choice (e.g. `app.py` in the Python tutorials)     
+│   └── 9. webservicve
+│       └── [The vulnerable application or files that you want to run/showcase using the TFW]    
+├── 10. config.yml exercise configuration file  
+└── 11. additional files (.circleci/; CODEOWNERS; README.md): should be there but   
+        nothing interesting for exercise development
+```
+
+### Attack
+
+Example file structure for an attack exercise:
+
+```   
+├── 1. controller  
+│   └── [solution checking]  
+├── 2. solvable  
+│   ├── 3. Dockerfile   
+│   └── 4. webservicve
+│       └── [The vulnerable application or files that you want to run/showcase]   
+├── 5. config.yml exercise configuration file  
+└── 6. additional files (.circleci/; CODEOWNERS; README.md): should be there but   
+        nothing interesting for exercise development 
+```
+
+## Detailed explanation
+
+### Controller
+
+This folder is for the solution checking. It can be:
+ * A flag, enabled in the `config.yml` file with `enable_flag_input: true`. You can specify a flag in the `config.yml` file like (like this: `flag: 'mySecretFlag'`), however, it is not recommended. Dynamic flags makes it harder to distribute the solution, it is advised to implement dynamic solution checking.
+ * A custom solution check if the `enable_flag_input` is set to `false`. The most common use case is to visit a page to check if the attack has been successful. You can use this [XSS injection solution check](https://github.com/avatao-content/challenge-toolbox/tree/v3/templates/xss/controller) as a good example.
+
+### Solvable
+
+The implemented application: the custom final state machine and event handler should be here.
+
+#### Solvable/Dockerfile
+
+Contains the commands to build up the solvable exercise's image. Please check the Dockerfile reference [here](https://docs.docker.com/engine/reference/builder/).
+
+Quick note: in the `VOLUME` array you can define the routes that’ll be excluded from the read-only file system. The basic commands of this file should be understood.
```

Comment:
Also, they'll be shared with other containers

Reply:
Thanks, I've added the extra info.